driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2282,6 +2285,195 @@ public void run() {
                 }
             }
         }
+
+        private class SchemaRefreshRequest {
+
+            private final String keyspace;
+
+            private final String table;
+
+            SchemaRefreshRequest(String keyspace, String table) {
+                this.keyspace = keyspace;
+                this.table = table;
+            }
+
+            /**
+             * Coalesce schema refresh requests.
+             * The algorithm is simple: if more than 2 keyspaces
+             * need refresh, then refresh the entire schema;
+             * otherwise if more than 2 tables in the same keyspace
+             * need refresh, then refresh the entire keyspace.
+             *
+             * @param that the other request to merge with the current one.
+             * @return A coalesced request
+             */
+            SchemaRefreshRequest coalesce(SchemaRefreshRequest that) {
+                if(this.keyspace == null || that.keyspace == null)
+                    return new SchemaRefreshRequest(null, null);
+                if(!this.keyspace.equals(that.keyspace))
+                    return new SchemaRefreshRequest(null, null);
+                if(this.table == null || that.table == null)
+                    return new SchemaRefreshRequest(keyspace, null);
+                if(!this.table.equals(that.table))
+                    return new SchemaRefreshRequest(keyspace, null);
+                return this;
+            }
+
+            @Override
+            public String toString() {
+                return Objects.toStringHelper(this)
+                    .add("keyspace", keyspace)
+                    .add("table", table)
+                    .toString();
+            }
+        }
+
+        private class SchemaRefreshRequestDeliveryCallback implements AbstractEventDebouncer.DeliveryCallback<SchemaRefreshRequest> {
+
+            @Override
+            public void deliver(final List<SchemaRefreshRequest> events) {
+                executor.submit(new ExceptionCatchingRunnable() {
```

Comment:
All the current 3 callbacks do the actual job on the main executor, never on the current thread (=scheduled executor thread or I/O thread).